src/ProtoRoute/Message.hs

```diff
@@ -1,46 +1,159 @@
+                       Invalid    String deriving (Show, Eq, Ord)
 
-data Field = Field
-    { fieldName :: FieldName
-    , fieldValue :: FieldValue
-    } deriving (Show)
+incomplete :: ValidationError
+incomplete = Incomplete "Required field not filled"
 
+incorrectFieldNames :: ValidationError
+incorrectFieldNames = WrongName "Wrong field names"
+
+incorrectMsgNames :: ValidationError
+incorrectMsgNames = WrongName "Message names do not match"
+
+incorrectTypes :: ValidationError
+incorrectTypes = WrongType "There are mismatching types"
+
+invalid :: ValidationError
+invalid = Invalid "Your message is invalid"
+
+-- Shows value of a message field
 fromFV :: FieldValue -> String
 fromFV (FText txt) = fromTV txt
 fromFV (FInt num)  = fromTV num
 fromFV (FMsg msg)  = fromTV msg
 
+-- Shows value inside FieldValue constructor
 fromTV :: (Show a) => TValue a -> String
 fromTV (Req val)  = show val
 fromTV (Opt txt)  = show txt
 fromTV (Rep list) = show list
 fromTV Error = "Error"
 
+-- Checks if rules in msg match those in corresponding schema
+correct :: TValue a -> SchemaRule -> Bool
+correct (Req _) SReq = True
+correct (Opt _) SOpt = True
+correct (Rep _) SRep = True
+correct _ _ = False
+
+-- Checks if types are correct
+okTypes :: FieldValue -> SchemaRule -> SchemaType -> Bool
+okTypes (FText tvA) rA fTA = fTA == SText && correct tvA rA
+okTypes (FInt tvB) rB fTB  = fTB == SInt && correct tvB rB
+okTypes (FMsg tvC) rC fTC  = fTC == SMsg && correct tvC rC
+
+-- Verifies if there exists a validation error
+verify :: Bool -> ValidationError -> Either ValidationError ()
+verify True someError = Left someError
+verify False _ = Right ()
+
+-- Checks if a required field is filled
+reqValThere :: SchemaField -> MessageField -> Bool
+reqValThere (SchemaField SReq _ sfn) mf =
+    case mf of
+        (MessageField sfn _) -> True
+reqValThere _ _ = False
+
+-- Validates single field, checking msg attempt against schema
+validateField :: SchemaField -> MessageField -> Either ValidationError ()
+validateField (SchemaField sfr sft sfn) (MessageField mfn mfv)  = do
+    verify (mfn /= sfn) incorrectFieldNames
+    verify (not $ okTypes mfv sfr sft) incorrectTypes
+    verify (not $ reqValThere (SchemaField sfr sft sfn) (MessageField mfn mfv))
+           incomplete
+
+thereExistsSF :: SchemaField -> Bool
+thereExistsSF (SchemaField _ _ name) = unFieldName name /= ""
+
+thereExistsMF :: MessageField -> Bool
+thereExistsMF (MessageField name _) = unFieldName name /= ""
+
+validateMessage :: MessageSchema -> Message -> Either ValidationError ()
+validateMessage (MessageSchema sn sFields) (Message mn mFields) = do
+    verify (mn /= sn) incorrectMsgNames
+    let nameFieldMap = groupedFields
+    let results = DMS.elems nameFieldMap
+    if null (catThese results)
+    then Left invalid
+    else Right ()
+  where
+    groupedFields :: DMS.Map FieldName (These SchemaField MessageField)
+    groupedFields = DMS.fromList $ zip fieldsNames theseFields
```

Comment:
Simple `zip` won't work here. Imagine the following two lists:

```haskell
schema = [("a", x), ("c", y)]
message = [("a", foo), ("b", bar), ("x", xxx)]
```

These lists should be grouped into the following map:

```haskell
"a" -> These x foo
"b" -> That bar
"c" -> This y
"x" -> That xxx
```

So you shouldn't use any assumptions on the positions of message fields. Intstead you should group by field name. I assume it would be simpler if you create these two maps from two lists:

```haskell
Map FieldName SchemaField
Map FieldName MessageField
```

And then you can combine both maps into 

```haskell
Map FieldName (These SchemaField MessageField)
```

There's programming pattern in Haskell: think in more generalized types. So I recommend you to implement the following pure function:

```haskell
unionThese :: Ord k => Map k a -> Map k b -> Map k (These a b)
```